### ISSUE TITLE
Alternative version / replacement of Replay Plugin 

### ISSUE THREAD

**Issue by AndreaCossu** [ContinualAI/avalanche#336]
Currently, Replay plugin keeps an external memory of patterns which are concatenated to the current step dataset. Minibatches are then sampled from this new dataset.

We could provide an additional Replay version or substitute the current one by concatenating to each _minibatch_ a random subset of the external memory. In this way, there is no need to concatenate to the entire dataset and replay is more effective (even if it costs more in terms of computation). 
This can be done in `before_forward` which is only applied to the training phase.

**Comment by Mattdl** [comment]
Replay is also not applicable in the data incremental learning setup when conditional on the class label. Even if the datastream is divided into experiences (e.g. 5 tasks MNIST), you want to check at every batch from which class a sample is, and based on this update the replay memory. Right now I don't see how I can add this to the dataset where we are already looping over.

**Comment by AndreaCossu** [comment]
Replay should be applicable also in data-incremental scenarios. I think it alread is in Avalanche. During a training experience, we select from the current training set at random (or with some more sophisticated policies) a subset of patterns. These patterns are added to the replay memory. When we want to do replay, we take some patterns from the replay memory and add it to the current training set, either by modifying the entire dataloader (as we do now) or by modifying the single minibatch (with `before_forward`, to be implemented). Does this clarify or have I completely misunderstood your point?

**Comment by Mattdl** [comment]
I see, we can indeed modify a single minibatch in `before_forward`, but then both (1) adding to the current minibatch
 and (2) updating the memory should happen there. However, we only have access to the actual inputdata (self.mb_x) which we would have to completely store for replay. It would be more efficient if we could just retain a dataset-idx or path to the img instead. (Which is done when concatenating the dataset)
 
Then, as you mentioned, altering the dataset is indeed not possible, as we are already looping over it (see https://github.com/ContinualAI/avalanche/blob/master/avalanche/training/strategies/base_strategy.py#L286 )

**Comment by AndreaCossu** [comment]
Why should we update the replay memory in the `before_forward`? We can do that right before the end of the experience by sampling from the current dataset. It only requires an additional pass over the data.

Also, if we concatenate replay patterns to each minibatch we don't need to alter the training dataset. We can sample some replay patterns and modifying the current minibatch by concatenating it with such patterns (basically, by modifying `self.mb_x` and `self.mb_y`).

**Comment by AntonioCarta** [comment]
> It would be more efficient if we could just retain a dataset-idx or path to the img instead

It is possible to make a specialized replay plugin that does this. However, saving the input data directly is more general and easier to implement. For example, if you save the dataset-idx you have to make sure to apply the same transformations. If you use the file path you are also assuming that your data is stored in a specific format.

Another alternative implementation may be to define a custom data loader that combines data from a specific experience with data from a buffer.

**Comment by Mattdl** [comment]
> Why should we update the replay memory in the before_forward? We can do that right before the end of the experience by sampling from the current dataset. It only requires an additional pass over the data.

Because then we assume that we know the experience/step/task boundaries right? But if we are really learning from a datastream, we don't know when an experience ends, and we would like to capture the exemplars after each batch we observe. So during the learning of the current experience, we might encounter exemplars from that experience already.

**Comment by Mattdl** [comment]
> > It would be more efficient if we could just retain a dataset-idx or path to the img instead
> 
> It is possible to make a specialized replay plugin that does this. However, saving the input data directly is more general and easier to implement. For example, if you save the dataset-idx you have to make sure to apply the same transformations. If you use the file path you are also assuming that your data is stored in a specific format.
> 
> Another alternative implementation may be to define a custom data loader that combines data from a specific experience with data from a buffer.

I think that would be interesting, certainly for larger-scale datasets where keeping buffers can become very memory demanding.

**Comment by AndreaCossu** [comment]
> > Why should we update the replay memory in the before_forward? We can do that right before the end of the experience by sampling from the current dataset. It only requires an additional pass over the data.
> 
> Because then we assume that we know the experience/step/task boundaries right? But if we are really learning from a datastream, we don't know when an experience ends, and we would like to capture the exemplars after each batch we observe. So during the learning of the current experience, we might encounter exemplars from that experience already.

Yes, at the moment we are assuming experience boundaries to be known at training time. You are right that if we remove this assumption things start to get complicated :grin:

**Comment by Mattdl** [comment]
> if we remove this assumption things start to get complicated

Or interesting (: Because it would enable us to include all of the online data incremental methods which learn from data streams (GSS, MIR, CoPE, CuRL, DNCPM, see https://arxiv.org/pdf/2009.00919.pdf for more info). The extension would only require to work on a batch-based setting instead of relying on the experience.

**Comment by AndreaCossu** [comment]
And interesting, for sure :) I'm only saying it would be an important (breaking) change. Removing experience affects all the codebase, not only replay.

**Comment by Mattdl** [comment]
I see, but I don't think we would have to remove experiences. The data stream can still be divided in experiences, it's just the methods should not take advantage or even be aware of the boundaries of experiences. (e.g. the real-life data stream can happen to be just like the typical Split-MNIST scenario)
So I think a simple solution with a custom dataloader like @AntonioCarta proposed, could already solve this issue (:

**Comment by vlomonaco** [comment]
Just a quick note: you can already split the experiences in datasets of 1 example each, so technically you could have a stream of data without "boundaries". Of course, that would be very inefficient in pytorch. I agree with the idea of fixing this strategies side!

**Comment by MattiaSangermano** [comment]
> We could provide an additional Replay version or substitute the current one by concatenating to each _minibatch_ a random subset of the external memory. In this way, there is no need to concatenate to the entire dataset and replay is more effective (even if it costs more in terms of computation).
> This can be done in `before_forward` which is only applied to the training phase.

Since the solution of putting the concatenation inside the `before_forward` would lead to using a number of examples in the mini-batch greater than those specified in the construction of the strategy and considering that a Dataloader should be used specifically for the memories, it is perhaps not better to replace the `self.adapted_dataset` with a Dataloader similar to the `MultiTaskMultiBatchDataLoader` which combines in each mini-batch examples from different experiences/memories?

**Comment by AntonioCarta** [comment]
yes. Keep in mind that `self.adapted_dataset` is the dataset, not the dataloader. You probably need to change the data-loader if you want to balance buffer and new data.

**Comment by MattiaSangermano** [comment]
Yes sorry, I meant `strategy.current_dataloader` not `strategy.adapted_dataset`
